src/main/java/org/prlprg/rds/RDSWriter.java

```diff
@@ -0,0 +1,552 @@
+      case GlobalEnvSXP _global -> RDSItemType.Special.GLOBALENV_SXP;
+      case SpecialSymSXP sexp when sexp == SEXPs.UNBOUND_VALUE ->
+          RDSItemType.Special.UNBOUNDVALUE_SXP;
+      case SpecialSymSXP sexp when sexp == SEXPs.MISSING_ARG -> RDSItemType.Special.MISSINGARG_SXP;
+
+        // BaseNamespace not supported
+
+      default -> new RDSItemType.Sexp(s.type());
+    };
+  }
+
+  // Returns the flags associated with the provided SEXP
+  private Flags flags(SEXP s) {
+    // FIXME: we should actually get the proper "locked" flag, but we currently don't have a
+    // representation of this in our environments
+    return new Flags(rdsType(s), new GPFlags(), s.isObject(), hasAttr(s), hasTag(s));
+  }
+
+  // Writes the provided flags to the stream
+  private void writeFlags(Flags flags) throws IOException {
+    out.writeInt(flags.encode());
+  }
+
+  // Writes the flags associated with the provided SEXP
+  private void writeFlags(SEXP s) throws IOException {
+    var flags = flags(s);
+    writeFlags(flags);
+  }
+
+  private void writeAttributes(Attributes attrs) throws IOException {
+    if (attrs.isEmpty())
+      throw new IllegalArgumentException("Cannot write an empty set of attributes");
+    // convert to ListSXP
+    var l = attrs.entrySet().stream().map(e -> new TaggedElem(e.getKey(), e.getValue())).toList();
+    // Write it
+    writeItem(SEXPs.list(l));
+  }
+
+  private void writeChars(String s) throws IOException {
+    // Never NA because we assume so
+    // We only consider scalar Na strings
+    var flags =
+        new Flags(
+            RDSItemType.valueOf(SEXPType.CHAR.i),
+            // FIXME: this should include the logic from platform.c (see rds-reader)
+            new GPFlags(Charset.defaultCharset(), false),
+            false,
+            false,
+            false);
+    writeFlags(flags);
+    out.writeInt(s.length());
+    out.writeString(s);
+  }
+
+  // SEXP writing ---------------------------------------------------------------------------------
+
+  private void writeEnv(EnvSXP env) throws IOException {
+    refAdd(env);
+    switch (env) {
+      case NamespaceEnvSXP namespace -> {
+        out.writeInt(RDSItemType.Special.NAMESPACESXP.i());
+        var namespaceInfo = SEXPs.string(namespace.name(), namespace.version());
+        write(namespaceInfo);
+        // TODO: needs tests
+      }
+      case UserEnvSXP userEnv -> {
+        out.writeInt(SEXPType.ENV.i);
+        // Write 1 if the environment is locked, or 0 if it is not
+        // FIXME: implement locked environments, as is this will always be false
+        out.writeInt(new GPFlags().isLocked() ? 1 : 0);
+        // Enclosure
+        writeItem(userEnv.parent());
+        // Frame
+        writeItem(userEnv.frame());
+        // Hashtab (NULL or VECSXP)
+        writeItem(SEXPs.NULL); // simple version here.
+        // Otherwise, we would have to actually do the hashing as it is done in R
```

Comment:
If you write tests on environment serialization that compare our version and R's, I guess we will not get the same output here (which is fine, but makes it more difficult to debug).